risc0/zkp/rust/src/verify/mod.rs

```diff
@@ -63,48 +62,154 @@ impl fmt::Display for VerificationError {
+                unsafe { core::slice::from_raw_parts_mut(out, len) }
+            } else {
+                panic!("Unable to convert between baby bear types &mut [{}] and &mut [{}]; other fields not supported yet.",
+                       core::any::type_name::<$from_ty>(),
+                       core::any::type_name::<$to_ty>());
+            }
+        }
+    };
+}
+
 pub trait VerifyHal {
     type Sha: Sha;
+    type Field: Field;
 
     fn sha(&self) -> &Self::Sha;
 
     fn debug(&self, msg: &str);
 
-    fn compute_polynomial(&self, u: &[Fp4], poly_mix: Fp4, out: &[Fp], mix: &[Fp]) -> Fp4;
+    fn compute_polynomial(
+        &self,
+        u: &[<Self::Field as Field>::ExtElem],
+        poly_mix: <Self::Field as Field>::ExtElem,
+        out: &[<Self::Field as Field>::Elem],
+        mix: &[<Self::Field as Field>::Elem],
+    ) -> <Self::Field as Field>::ExtElem;
 
     /// Evaluate a polynomial whose coefficients are in the extension field at a
     /// point.
-    fn poly_eval(&self, coeffs: &[Fp4], x: Fp4, y: Fp) -> Fp4 {
-        let mut mul_fp = Fp::ONE;
-        let mut mul_fp4 = Fp4::ONE;
-        let mut tot = Fp4::ZERO;
+    fn poly_eval(
+        &self,
+        coeffs: &[<Self::Field as Field>::ExtElem],
+        x: <Self::Field as Field>::ExtElem,
+        y: <Self::Field as Field>::Elem,
+    ) -> <Self::Field as Field>::ExtElem {
+        let mut mul_fp = <Self::Field as Field>::Elem::ONE;
+        let mut mul_fp4 = <Self::Field as Field>::ExtElem::ONE;
+        let mut tot = <Self::Field as Field>::ExtElem::ZERO;
         for i in 0..coeffs.len() {
             tot += coeffs[i] * mul_fp * mul_fp4;
             mul_fp *= y;
             mul_fp4 *= x;
         }
         tot
     }
+
+    // Adapters to convert to/from baby bear field until all code is migrated.
+    // TODO: Remove once migration complete
+    baby_bear_adapter!(
+        to_baby_bear_fp,
+        to_baby_bear_fp_slice,
+        to_baby_bear_fp_slice_mut,
+        <Self::Field as Field>::Elem,
+        <BabyBear as Field>::Elem,
+    );
+    baby_bear_adapter!(
+        to_baby_bear_fp4,
+        to_baby_bear_fp4_slice,
+        to_baby_bear_fp4_slice_mut,
+        <Self::Field as Field>::ExtElem,
+        <BabyBear as Field>::ExtElem,
+    );
+    baby_bear_adapter!(
+        from_baby_bear_fp,
+        from_baby_bear_fp_slice,
+        from_baby_bear_fp_slice_mut,
+        <BabyBear as Field>::Elem,
+        <Self::Field as Field>::Elem,
+    );
+    baby_bear_adapter!(
+        from_baby_bear_fp4,
+        from_baby_bear_fp4_slice,
+        from_baby_bear_fp4_slice_mut,
+        <BabyBear as Field>::ExtElem,
+        <Self::Field as Field>::ExtElem,
+    );
 }
 
 #[cfg(feature = "host")]
 mod host {
     use super::*;
     use crate::adapter::{PolyExt, PolyExtContext};
 
-    pub struct CpuVerifyHal<'a, S: Sha, C: PolyExt> {
+    // TODO: Not certain how best to handle the unused Field that's needed to
```

Comment:
The `CpuVerifyHal` needs to know which `Field` elements it will be working with, but it doesn't actually directly use the field -- is this the right design for providing that information?

Reply:
I think you can use `PhantomData` for this purpose.

https://doc.rust-lang.org/std/marker/struct.PhantomData.html